### PR TITLE
Change IMessageWithContentType to MessageWithMetadata

### PR DESCRIPTION
**Pull request by JoshLove-msft**

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs [hunk 1]
```diff
@@ -42,10 +42,17 @@ public class EventData
         /// <seealso cref="BinaryData" />
         /// <seealso cref="EventData.Properties" />
         ///
-        public BinaryData EventBody
+        public BinaryData EventBody => _amqpMessage.GetEventBody();
```
Comment (JoshLove-msft) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs:45]: public setter is no longer needed (this was only shipped in a beta)

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs [hunk 2]
```diff
@@ -428,14 +435,6 @@ public EventData(BinaryData eventBody) : this(eventBody, lastPartitionSequenceNu
         {
         }
 
-        /// <summary>
-        ///   Initializes a new instance of the <see cref="EventData"/> class.
-        /// </summary>
-        ///
-        public EventData() : this(new BinaryData(Array.Empty<byte>()), lastPartitionSequenceNumber: null)
```
Comment (JoshLove-msft) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs:435]: No longer needed (only shipped in a beta)

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md [hunk 3]
```diff
@@ -82,30 +76,8 @@ The following shows examples of what is available through the SchemaRegistryAvro
...
-
-### Serialize
-
-Register a schema to be stored in the Azure Schema Registry.
-
-```C# Snippet:SchemaRegistryAvroSerialize
-var employee = new Employee { Age = 42, Name = "John Doe" };
+### Encode and decode data using the Event Hub EventData model
 
-using var memoryStream = new MemoryStream();
```
Comment (JoshLove-msft) [inline sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md:95]: Need to add new snippets

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/api/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro.netstandard2.0.cs [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
+        public object DecodeMessageData(Azure.Messaging.MessageWithMetadata message, System.Type returnType, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public System.Threading.Tasks.ValueTask<object> DecodeMessageDataAsync(Azure.Messaging.MessageWithMetadata message, System.Type returnType, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public System.Threading.Tasks.ValueTask<T> EncodeMessageDataAsync<T>(object data, System.Type inputType = null, System.Func<System.BinaryData, T> func = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) where T : Azure.Messaging.MessageWithMetadata { throw null; }
+        public T EncodeMessageData<T>(object data, System.Type inputType = null, System.Func<System.BinaryData, T> func = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) where T : Azure.Messaging.MessageWithMetadata { throw null; }
```
Comment (JoshLove-msft) [inline sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/api/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro.netstandard2.0.cs:9]: rename `func` to something like messageFactory

File: sdk/eventhub/Azure.Messaging.EventHubs.Shared/tests/Azure.Messaging.EventHubs.Shared.Tests.csproj [hunk 5]
```diff
@@ -27,6 +31,7 @@
   <ItemGroup>
     <ProjectReference Include="../../Azure.Messaging.EventHubs/src/Azure.Messaging.EventHubs.csproj" />
     <ProjectReference Include="$(AzureCoreTestFramework)" />
+    <ProjectReference Include="..\..\..\core\Azure.Core\src\Azure.Core.csproj" />
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Shared/tests/Azure.Messaging.EventHubs.Shared.Tests.csproj:34]: nit:  Do we need the explicit reference?   

_(I'm not sure why we've got one at all, honestly.  We should pick it up transiently via the Event Hubs project reference.  Whatever reason that we had, I can't remember it.)_

**Review by jsquire — APPROVED** [review]
LGTM; I left a couple of small formatting nits and wanted to consider leaving the body setter for events.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs:45]: I don't mind keeping it - your call 😄

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs:45]: cool.   Let's keep them then.

File: sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs [hunk 6]
```diff
@@ -115,3 +115,12 @@ public static partial class ResponseExtensions
         public static bool IsError(this Azure.Response response) { throw null; }
     }
 }
+namespace Azure.Messaging
+{
+    public abstract partial class MessageWithMetadata
+    {
+        protected MessageWithMetadata() { }
+        public abstract string ContentType { get; set; }
+        public abstract System.BinaryData Data { get; set; }
```
Comment (KrzysztofCwalina) [inline sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: So, are we adding IsReadOnly property?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: No I don't think it is necessary at this point. If a read-only message is passed to an API that needs to modify it, we have no choice but to throw.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: We can make the "throwing" a responsibility of the encoder by adding IsReadOnly, but we would still need to throw from the setters in the read-only types, so I didn't see a benefit to adding it.

Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: It's a common established pattern in .NET to have abstractions that can be read-write or read-only (e.g. IList). Such abstraction must expose IsReadOnly property such that calling code can avoid exceptions. Reversely, it's not idiomatic to simply throw from such methods.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: In this case, I don't see how we can avoid throwing an exception if a ServiceBusReceivedMessage is passed into the Encode method.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:124]: I suppose the consumer can check the IsReadOnly property and elect to not call Encode if IsReadOnly is true, but I don't see this being a common scenario.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.netstandard2.0.cs:103]: I suppose this isn't something that we can add later as it needs to be abstract. I will add it given that this is the established convention.

**Review by KrzysztofCwalina — APPROVED** [review]

**Pull request merged**